grafana/resource_stack.go

```diff
@@ -0,0 +1,274 @@
+package grafana
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"path"
+	"strconv"
+	"strings"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func ResourceStack() *schema.Resource {
+	return &schema.Resource{
+
+		Description: `
+* [Official documentation](https://grafana.com/docs/grafana-cloud/reference/cloud-api/#stacks/)
+* [HTTP API](https://grafana.com/docs/grafana/latest/http_api/stack/)
+`,
+
+		CreateContext: CreateStack,
+		UpdateContext: UpdateStack,
+		DeleteContext: DeleteStack,
+		Exists:        ExistsStack,
+		ReadContext:   ReadStack,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The stack id assigned to this stack by Grafana.",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of stack. Conventionally matches the url of the instance (e.g. “<stack_slug>.grafana.net”).",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of stack.",
+			},
+			"slug": {
+				Type:     schema.TypeString,
+				Required: true,
+				Description: `
+Subdomain that the Grafana instance will be available at (i.e. setting slug to “<stack_slug>” will make the instance
+available at “https://<stack_slug>.grafana.net".`,
+			},
```

Comment:
Since slugs can only support alphanumeric chars, you can have a validation at plan time

```suggestion
var stackSlugRegex = regexp.MustCompile("^[a-z0-9]+$")

.....

			"slug": {
				Type:     schema.TypeString,
				Required: true,
				Description: `
Subdomain that the Grafana instance will be available at (i.e. setting slug to “<stack_slug>” will make the instance
available at “https://<stack_slug>.grafana.net".`,
			},
			"slug": {
				Type:     schema.TypeString,
				Required: true,
				Description: `
Subdomain that the Grafana instance will be available at (i.e. setting slug to “<stack_slug>” will make the instance
available at “https://<stack_slug>.grafana.net".`,
				ValidateFunc: validation.StringMatch(stackSlugRegex, "must be a lowercase alphanumeric string"),
			},
```